sender.py

```diff
@@ -0,0 +1,150 @@
+#! /usr/bin/env python3
+
+import logging
+import os
+import smtplib
+import socket
+import time
+
+import dns.resolver
+
+import mailqueue
+
+def main():
+    setup_logging()
+
+    delivery_agent = DeliveryAgent(mailqueue.MailQueue(), get_dns_resolver(), SMTPClient())
+
+    if 'RETRY_INTERVAL' in os.environ:
+        delivery_agent.retry_interval = int(os.environ['RETRY_INTERVAL'])
+
+    while True:
+        # TODO: retrieve a batch of messages and group by destination domain to reuse SMTP sessions
+        # to the same MX
+        if delivery_agent.deliver_single_envelope() == DeliveryAgent.IDLE:
+            time.sleep(1)
+
+
+class DeliveryAgent:
+    IDLE = 'IDLE (there was no items in the incoming queue)'
+    DONE = 'DONE (there was at least one item to process)'
+
+    def __init__(self, mailqueue, dns_resolver, smtp_client):
+        self.max_delivery_attempts = 4 # 1 initial, 3 retries
+        self.retry_interval = 600 # seconds
+
+        self._dns_resolver = dns_resolver
+        self._mailqueue = mailqueue
+        self._smtp_client = smtp_client
+
+    def deliver_single_envelope(self):
+        envelope = self._mailqueue.get()
```

Comment:
If you're still looking for last minute requests I'd ask for scaling senders.

Reply:
I'm currently working on making the sender multi-threaded. This will help with scale, but we can go further be running multiple `sender` processes (this will require sharding).

Do you want me to work on that before completing other items in the TODO?

Reply:
If multi-threaded sender is close to completion I'd finish it first. Curious how it's gonna look like. And I'd do the multiple sender processes next.

Reply:
@obukhov-sergey, implemented both multi-threading and sharded senders, please take a look.